# Урок 2

## Дискретные случайные величины. Закон распределения вероятностей.  Биномиальный закон распределения. Распределение Пуассона

In [1]:
import numpy as np
from math import factorial
from math import exp
import warnings
warnings.filterwarnings('ignore')

### Задача 1

#### Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

1. Результат ответа на любой из вопросов не зависит от ответов на другие вопросы, т.е. события независимые.
2. Вероятность правильного ответа на любой из вопросов равна $\dfrac{1}{4}$ (0.25)
3. Как нам хорошо известно, биномиальный закон описывает распределение случайной величины 𝑋, задающей число наступлений события  𝐴 в ходе проведения 𝑛 независимых опытов.

__Т.о., мы можем воспользоваться основными вывадами (формулами) из биномиального закона.__

$M(X) = np\; = 5 \cdot \dfrac{1}{4} = 1.25\\ D(X) = np(1 - p)\; = 5 \cdot \dfrac{1}{4}\cdot(1-\dfrac{1}{4})\; = 0.9375 $

Возможные значения случайной величины (в нашем случае правильных ответов) при простом угадывании: $x_1 = 0$, $x_2 = 1$, $x_3 = 2$, $x_4 = 3$, $x_5 = 4$, $x_6 = 5$. Посчитаем вероятность каждого значения, используя __формулу Бернулли__ $P_n = C_n^k p^k (1 - p)^{n - k}\;$:

In [2]:
def combinations(n: int, k: int) -> int:
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

distr = []

for i in range(0, 6):
    P = np.round(combinations(5, i)*.25**i*.75**(5-i),4)
    print(f'P({i})={P}')
    distr.append(P)

P(0)=0.2373
P(1)=0.3955
P(2)=0.2637
P(3)=0.0879
P(4)=0.0146
P(5)=0.001


Итак, закон распределения такой случайной величины:

<table border="3">
    <tr>
        <th>$x$</th>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
        <td>5</td>
    </tr>
    <tr>
        <th>$P(X = x)$</th>
        <td>0.2373</td>
        <td>0.3955</td>
        <td>0.2637</td>
        <td>0.0879</td>
        <td>0.0146</td>
        <td>0.001</td>
    </tr>
</table>

### Задача 2

#### Вероятность попадания в цель при одном выстреле равна 0.01. Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

Похоже мы имеем дело со счетчиком событий с достаточно большим количеством испытаний и достаточно малой вероятностью благоприятных исходов. Поэтому для решения данной задачи воспользуемся формулой Пуассона: $P(X = k) = \dfrac{a^k e ^{-a}}{k!}\; \approx\large \frac{\lambda^m}{m!}e^{-\lambda},\; где\; λ = np = 200 \cdot 0.01 = 2$.

По теореме сложения вероятностей имеем:

$P(5<=X<=10) = P(X=5) + P(X=6) + P(X=7) + P(X=8) + P(X=9) + P(X=10)$

In [3]:
p, n = 0.01, 200

def puasson(n, m, p):
    lambda_ = n * p
    return (lambda_**m / np.math.factorial(m)) * np.exp(-lambda_)

sum_ = 0
for i in range(5, 11):
    sum_ += puasson(n, i, p)

print(f'Вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10 =', np.round(sum_,4))

Вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10 = 0.0526


### Задача 3

#### В условиях предыдущей задачи сколько выстрелов нужно сделать, чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание?

#### Подсказка. Как посчитать P(X >= k) для величины X, имеющей распределение Пуассона? Здесь предстоит немножко покодить.

По сути нужно найти алгоритм со счетчиком и циклом, который перебирал бы значения n до тех пор, пока сумма вероятностей не достигнет искомого значения. Кажется нужно искать не p, а q.

In [4]:
p, k, n = 0.01, 10, 10

def puss(n, k, p):
    sum_ = 0
    for i in range(0, k):  
        sum_ += puasson(n, i, p)
    return 1 - sum_

while puss(n, k, p) <= 0.9:
    n += 1
    
print(f'Нужно сделать {n} выстрелов, чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание.\n'
      f'На всякий случай: вероятность равна {np.round(puss(n, k, p),4)}')

Нужно сделать 1421 выстрелов, чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание.
На всякий случай: вероятность равна 0.9002
